dali/operators/decoder/audio/generic_decoder.cc

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/decoder/audio/generic_decoder.h"
+
+namespace dali {
+
+namespace {
+
+template<typename SampleType>
+size_t read_samples(SNDFILE *snd_file, span<SampleType> output);
+//  DALI_FAIL("Can't find function for given type. You may want to write your own specialization.")
+//}
+
+
+template<>
+size_t read_samples<short>(SNDFILE *snd_file, span<short> output) {  // NOLINT
+  return sf_read_short(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<int>(SNDFILE *snd_file, span<int> output) {
+  return sf_read_int(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<float>(SNDFILE *snd_file, span<float> output) {
+  return sf_read_float(snd_file, output.data(), output.size());
+}
+
+
+template<>
+size_t read_samples<double>(SNDFILE *snd_file, span<double> output) {
+  return sf_read_double(snd_file, output.data(), output.size());
+}
+
+
+class MemoryStream {
+ public:
+  MemoryStream() : length_(0), curr_(0), input_(nullptr) {}
```

Comment:
Why not express a default constructor as parametrized one with 0 and nullptr as the arguments. In that way you need to initialize variables in one place instead of two.

Reply:
Done